dockerHashing/src/main/scala/cromwell/docker/registryv2/flows/azure/AzureContainerRegistry.scala

```diff
@@ -0,0 +1,149 @@
+package cromwell.docker.registryv2.flows.azure
+
+import cats.data.Validated.{Invalid, Valid}
+import cats.effect.IO
+import com.typesafe.scalalogging.LazyLogging
+import common.validation.ErrorOr.ErrorOr
+import cromwell.cloudsupport.azure.AzureCredentials
+import cromwell.docker.DockerInfoActor.DockerInfoContext
+import cromwell.docker.{DockerImageIdentifier, DockerRegistryConfig}
+import cromwell.docker.registryv2.DockerRegistryV2Abstract
+import org.http4s.{Header, Request, Response, Status}
+import cromwell.docker.registryv2.flows.azure.AzureContainerRegistry.domain
+import org.http4s.circe.jsonOf
+import org.http4s.client.Client
+import io.circe.generic.auto._
+import org.http4s._
+
+
+class AzureContainerRegistry(config: DockerRegistryConfig) extends DockerRegistryV2Abstract(config) with LazyLogging {
+
+  /**
+    * (e.g registry-1.docker.io)
+    */
+  override protected def registryHostName(dockerImageIdentifier: DockerImageIdentifier): String =
+    dockerImageIdentifier.host.getOrElse("")
+
+  override def accepts(dockerImageIdentifier: DockerImageIdentifier): Boolean =
+    dockerImageIdentifier.hostAsString.contains(domain) 
+
+  override protected def authorizationServerHostName(dockerImageIdentifier: DockerImageIdentifier): String =
+    dockerImageIdentifier.host.getOrElse("")
+
+  /**
+    * In Azure, service name does not exist at the registry level, it varies per repo, e.g. `terrabatchdev.azurecr.io`
+    */
+  override def serviceName: Option[String] =
+    throw new Exception("ACR service name is host of user-defined registry, must derive from `DockerImageIdentifier`")
+  
+  /**
+    * Builds the list of headers for the token request
+    */
+  override protected def buildTokenRequestHeaders(dockerInfoContext: DockerInfoContext): List[Header] = {
+    List(contentTypeHeader)
+  }
+
+  private val contentTypeHeader: Header = {
+    import org.http4s.headers.`Content-Type`
+    import org.http4s.MediaType
```

Comment:
Any particular reason for having these imports embedded in these functions rather than at the top?

Reply:
Personal preference really, they are extremely specific and only used in one place.